nix/grafana/default.nix

```diff
@@ -0,0 +1,83 @@
+{ pkgs, lib, name, config, ... }:
+let
+  inherit (lib) types;
+  iniFormat = pkgs.formats.ini { };
+in
+{
+  options = {
+    description = ''
+      Configure grafana.
+    '';
+    enable = lib.mkEnableOption name;
+
+    package = lib.mkPackageOption pkgs "grafana" { };
+
+    port = lib.mkOption {
+      type = types.int;
+      description = "Which port to run grafana on.";
+      default = 3000;
+    };
+
+    dataDir = lib.mkOption {
+      type = types.str;
+      description = "Directory where grafana stores its logs and data.";
+      default = "./data/${name}";
+    };
+
+    extraConf = lib.mkOption {
+      type = iniFormat.type;
+      description = "Extra configuration for grafana.";
+      default = { };
+      example = ''
+        {
+          security.admin_user = "patato";
+          security.admin_password = "potato";
+        }
+      '';
+    };
+
+    outputs.settings = lib.mkOption {
+      type = types.deferredModule;
+      internal = true;
+      readOnly = true;
+      default = {
+        processes."${name}" =
+          let
+            grafanaConfig =
+              {
+                server = {
+                  protocol = "http";
+                  http_port = config.port;
+                  domain = "localhost";
+                };
+              } // config.extraConf;
+            grafanaConfigIni = iniFormat.generate "defaults.ini" grafanaConfig;
+            startScript = pkgs.writeShellApplication {
+              name = "start-grafana";
+              runtimeInputs = [ config.package ];
+              text = ''
+                grafana server --config ${grafanaConfigIni} \
+                               --homepath ${config.package}/share/grafana \
+                               cfg:paths.data="$(readlink -m ${config.dataDir})"
+              '';
+            };
+          in
+          {
+            command = startScript;
+            readiness_probe = {
+              exec.command = "${pkgs.curl}/bin/curl -f ${grafanaConfig.server.protocol}://${grafanaConfig.server.domain}:${builtins.toString grafanaConfig.server.http_port}/api/health";
```

Comment:
You can use the [http_get](https://github.com/Platonic-Systems/process-compose-flake/blob/b346fb30ad5fdee14e3edc3ba490eaa5db9ec0b7/nix/process-compose/settings/probe.nix#L13-L37) readiness probe instead of curl.

See example: https://github.com/Platonic-Systems/process-compose-flake/blob/b346fb30ad5fdee14e3edc3ba490eaa5db9ec0b7/example/flake.nix#L56-L59

Reply:
Also, perhaps make the whole `${grafanaConfig.server.protocol}://${grafanaConfig.server.domain}:${builtins.toString grafanaConfig.server.http_port}"` a value of an internal option (`grafanaUri`?), so it can be re-used in other places, including by the user. It also makes the probe CLI here more readable.

Reply:
@srid sure.

grafana has a config called `root_url` for this. i'll store the resulting value in this config